### PR TITLE
Feat: Add Bundler and tighten Docker Image

### PR DESCRIPTION
**Pull request by ardelato**
## Description

This has some major changes to the build logic and dev environment. The core logic remains the same.

### Changes:
- Restructured the files in the repo to be more granular with the content
- Added `tsup` and `tsc-alias` to compile and bundle the code for specific uses and targets
- Added `path` aliases in the `tsconfig.json`
- Added new helper scripts to get Devs setup quickly and run docker

### CR

The main goal of these changes were to allow us to have individual folders for our Typescript code -- i.e. config, core, and scripts. Then when we needed to execute the code, it would be properly compiled without nested folders in the `dist` folder. To accomplish that, we needed a bundler that could parse all the Typescript code and compile it for Node as standalone JS files.

I tried out different bundlers: `rspack`, `swc`, and `webpack`. They all had their own issues and I couldn't get them to work. This was mainly due to a Node issue with ES modules and CJS modules not being able to be mixed but also not being able to be separated when working on a Node app:
https://github.com/evanw/esbuild/issues/1921#issuecomment-1784828307.

At the end of the day I only got `esbuild` to work and in the desired way to address both #11 and #14. That said, `esbuild` still had its own issues which led me to discover the tool [`tsup`](https://tsup.egoist.dev). `tsup` is a wrapper around `esbuild` and fixes some of the underlying issues of working with the `esbuild` API including shimming modules -- https://tsup.egoist.dev/#inject-cjs-and-esm-shims. 

Aside from that, we also need the ability to not include the config files -- i.e `lh-config.js` and `urls.json` -- in the bundled `dist/index.js` file in order to manually configure these options for the Docker container. If they were included then we would have to rebuild the Docker image every time we wanted to change the configs and would need to build and tag multiple images for variations in the configs. d9d6942775db809cd28249de74ca6411abadf8d0 provides further details and explanation around this requirement.

### QA

- [ ] The `setup` script works
	1. Run `pnpm run setup`
	2. Confirm the prior templated files (`.env`, `lh-config.js`, and `urls.json`) have been copied to the appropriate locations.
- [ ] The `run-docker` script works
	1. Follow the `README` setup instructions
	2. Build the docker image `docker build -t vigilo .`
	3. Run `pnpm run start:docker`
- [ ] Confirm the `.env` file is not copied over to the docker image
	1. Follow the `README` setup instructions
	2. Build the docker image `docker build -t vigilo .`
	3. Run `docker run --rm -it vigilo /bin/bash`
	4. `ls` to ensure the `.env` file was not copied
- [ ] `start` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/1
- [ ] `updateDatadogMetricsMetadata` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/9
- [ ] `createDashboards` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/8

Closes: #11
Closes: #14 

cr_req 2

### CODE REVIEW

**Comment by deltuh-vee** [comment]
dev_block ✊ 
Ran into this error when trying to run `pnpm run start`.
CleanShot 2023-12-01 at 11 08 03

**Comment by ardelato** [comment]
un_dev_block ☁️ , @deltuh-vee try running it with `node v18`. There is a warning message at the top saying `node v16` was used

**Comment by deltuh-vee** [comment]
I get a different error now. Does dist/index.js still exist?
CleanShot 2023-12-01 at 14 54 23

**Comment by deltuh-vee** [comment]
Was index.js supposed to be moved out of the src directory? I still see it there.
CleanShot 2023-12-01 at 14 59 20

**Comment by deltuh-vee** [comment]
QA 🎬
- `pnpm run setup` copies .env to the root directory while lh-config.js, and urls.json are copied to the config folder
- The `run-docker` script works
- The .env file is not copied over to the docker image
- The start, updateDatadogMetricsMetadata, createDashboards scripts still work.

File: dev-scripts/setup.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+set -ex
+
+root_dir=$(git rev-parse --show-toplevel)
+
+config_templates_dir="$root_dir/config-templates"
+config_dir="$root_dir/src/config"
+
+cp $config_templates_dir/.env.template $root_dir/.env
```
Comment (batbattur) [inline dev-scripts/setup.sh:10]: Question: Should the `.env` file for Datadog config be in `config-templates` or the root dir? If it's supposed to be in the root dir should we move the `.env.template` file to the root dir?

**Review by batbattur — APPROVED** [review]
CR 🚛 makes sense to me. Nice work @ardelato!

I'm going to set cr req 2 since I'm not very familiar with bundlers

File: dev-scripts/setup.sh (hunk 1 above)
Comment (ardelato) [reply dev-scripts/setup.sh:10]: I created a `config-templates` directory to house all the configuration files that are mainly boilerplate files. The locations of where they should end up are somewhat enforced by the import linters and paths aliased in `tsconfig.json`. The setup script just helps automate the process of copying the boilerplates to the expected locations.

Think of the `config-templates` folder like the `templates` folder for our `Ansible` roles in `server-templates`

**Comment by batbattur** [comment]
Sorry, accidentally pushed [unrelated commit](https://github.com/iFixit/vigilo/commit/c58a9d6fd508c52cb0a02d68595898b2f808521c) to the branch! Reverted to the last commit. 

CR 🚛  from https://github.com/iFixit/vigilo/pull/17#pullrequestreview-1762982063
QA 🛻 from https://github.com/iFixit/vigilo/pull/17#issuecomment-1836919999

**Review by mlahargou — APPROVED** [review]
CR 📱

**Pull request merged**